zk-kyc/zkvm/host/src/lib.rs

```diff
@@ -0,0 +1,55 @@
+pub mod seal_serializer {
+    //! Module for serializing seal structures
+
+    use bonsai_sdk::alpha::responses::Groth16Seal;
+    use hex;
+    use serde::ser::SerializeStruct;
+    use serde::{Serialize, Serializer};
+
+    /// Serializes a vector of bytes as a hexadecimal string.
+    /// Returns a result containing either the serialized object or an error.
+    pub fn serialize_as_hex<S>(bytes: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&hex::encode(bytes))
+    }
+
+    /// Serializes a `Groth16Seal` struct, encoding its fields as hexadecimal strings.
+    /// Returns a result containing either the serialized object or an error.
+    pub fn serialize_groth16_seal<S>(groth: &Groth16Seal, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("Groth16Seal", 4)?;
+        s.serialize_field("a", &groth.a.iter().map(hex::encode).collect::<Vec<_>>())?;
+        s.serialize_field(
+            "b",
+            &groth
+                .b
+                .iter()
+                .map(|inner_vec| inner_vec.iter().map(hex::encode).collect::<Vec<_>>())
+                .collect::<Vec<_>>(),
+        )?;
+        s.serialize_field("c", &groth.c.iter().map(hex::encode).collect::<Vec<_>>())?;
+        s.serialize_field(
+            "public",
+            &groth.public.iter().map(hex::encode).collect::<Vec<_>>(),
+        )?;
+        s.end()
+    }
+
+    /// A wrapper struct for SnarkReceipts.
+    ///
+    /// This struct is used to serialize a `Groth16Seal` along with its associated
+    /// `post_state_digest` and `journal`, both of which are serialized as hexadecimal strings.
+    #[derive(Serialize)]
+    pub struct SnarkReceiptWrapper<'a> {
+        #[serde(serialize_with = "serialize_groth16_seal")]
+        pub snark: &'a Groth16Seal,
+        #[serde(serialize_with = "serialize_as_hex")]
+        pub post_state_digest: &'a Vec<u8>,
+        #[serde(serialize_with = "serialize_as_hex")]
+        pub journal: &'a Vec<u8>,
```

Comment:
Optionally, hex has a util for this:
```diff
diff --git a/zk-kyc/zkvm/host/Cargo.toml b/zk-kyc/zkvm/host/Cargo.toml
index cfc7b9f..36a95d9 100644
--- a/zk-kyc/zkvm/host/Cargo.toml
+++ b/zk-kyc/zkvm/host/Cargo.toml
@@ -13,7 +13,7 @@ clap = { version = "=4.3.23", features = ["derive"] }
 cookie = "0.18.0"
 dotenv = "0.15.0"
 futures = "0.3.28"
-hex = "0.4.3"
+hex = { version = "0.4.3", features = ["serde"] }
 jwt-compact = { version = "0.7.0", features = ["rsa"] }
 log = "0.4.20"
 methods = { path = "../methods" }
diff --git a/zk-kyc/zkvm/host/src/lib.rs b/zk-kyc/zkvm/host/src/lib.rs
index 0e6283d..cbd815d 100644
--- a/zk-kyc/zkvm/host/src/lib.rs
+++ b/zk-kyc/zkvm/host/src/lib.rs
@@ -6,15 +6,6 @@ pub mod seal_serializer {
     use serde::ser::SerializeStruct;
     use serde::{Serialize, Serializer};
 
-    /// Serializes a vector of bytes as a hexadecimal string.
-    /// Returns a result containing either the serialized object or an error.
-    pub fn serialize_as_hex<S>(bytes: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_str(&hex::encode(bytes))
-    }
-
     /// Serializes a `Groth16Seal` struct, encoding its fields as hexadecimal strings.
     /// Returns a result containing either the serialized object or an error.
     pub fn serialize_groth16_seal<S>(groth: &Groth16Seal, serializer: S) -> Result<S::Ok, S::Error>
@@ -47,9 +38,9 @@ pub mod seal_serializer {
     pub struct SnarkReceiptWrapper<'a> {
         #[serde(serialize_with = "serialize_groth16_seal")]
         pub snark: &'a Groth16Seal,
-        #[serde(serialize_with = "serialize_as_hex")]
+        #[serde(serialize_with = "hex::serialize")]
         pub post_state_digest: &'a Vec<u8>,
-        #[serde(serialize_with = "serialize_as_hex")]
+        #[serde(serialize_with = "hex::serialize")]
         pub journal: &'a Vec<u8>,
     }
 }

```

Reply:
I'm curious - how did you create this snippet with git? Its very nice :D